ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/validate/ConflictCheckWithCommittedImpl.scala

```diff
@@ -134,54 +135,88 @@ private[validate] class ConflictCheckWithCommittedImpl(
-            }
+  ): EitherT[Future, Rejection, Unit] =
+    keyInputs.toList.collectFirstSomeM { case (key, inputState) =>
+      indexService
+        .lookupContractKey(transactionInformees, key)(loggingContext)
+        .map { lookupResult =>
+          (inputState, lookupResult) match {
+            case (LfTransaction.NegativeKeyLookup, Some(actual)) =>
+              Left(InconsistentContractKey(None, Some(actual))(completionInfo))
+            case (LfTransaction.KeyCreate, Some(_)) =>
+              Left(DuplicateKey(key)(completionInfo))
+            case (LfTransaction.KeyActive(expected), actual) if !actual.contains(expected) =>
+              Left(InconsistentContractKey(Some(expected), actual)(completionInfo))
+            case _ => Right(())
+          }
         }
-    }
-  }
+        .map(_.left.toOption)
+    }.invertToEitherT
 
   private def validateExplicitDisclosure(
       disclosedContracts: ImmArray[Versioned[DisclosedContract]],
       completionInfo: CompletionInfo,
   )(implicit
       contextualizedErrorLogger: ContextualizedErrorLogger,
       loggingContext: LoggingContext,
-  ): EitherT[Future, Rejection, Unit] = {
-    // Note that the validation fails fast on the first unknown/invalid contract.
-    disclosedContracts.foldLeft(EitherT(Future.successful[Validation[Unit]](Right(())))) {
-      case (f, provided) =>
-        f.flatMapF { _ =>
-          indexService
-            .lookupContractStateWithoutDivulgence(provided.unversioned.contractId)
-            .map {
-              case ContractState.Archived | ContractState.NotFound =>
-                // Disclosed contract was archived or never existed
-                Left(UnknownContracts(Set(provided.unversioned.contractId))(completionInfo))
-              case ContractState.Active(contractInstance, ledgerEffectiveTime) =>
-                sameContractData(contractInstance, ledgerEffectiveTime, provided).left.map {
-                  errMessage =>
-                    logger.info(errMessage)
-                    DisclosedContractInvalid(provided.unversioned.contractId, completionInfo)
-                }
+  ): EitherT[Future, Rejection, Unit] =
+    // Validation fails fast on the first unknown/invalid contract.
+    disclosedContracts.toList.collectFirstSomeM { disclosedContract =>
+      val validation = for {
+        _ <- validateDisclosedContractPayload(disclosedContract, completionInfo)
+        _ <- validateDriverMetadataContractId(completionInfo, disclosedContract)
+      } yield ()
+
+      validation.value.map(_.left.toOption)
+    }.invertToEitherT
+
+  private def validateDisclosedContractPayload(
+      provided: Versioned[DisclosedContract],
+      completionInfo: CompletionInfo,
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger,
+      loggingContext: LoggingContext,
+  ): EitherT[Future, Rejection, Unit] =
+    EitherT(
+      indexService
+        .lookupContractStateWithoutDivulgence(provided.unversioned.contractId)
```

Comment:
Nit, can you assign `provided.unversioned.contractId` to val? as it appears three times in this method; some for the method below